pythonforandroid/bootstraps/webview/build/src/main/java/org/kivy/android/PythonActivity.java

```diff
@@ -173,13 +185,55 @@ public void onClick(DialogInterface dialog,int id) {
         mWebView.loadUrl("file:///" + app_root_dir + "/_load.html");
 
         mWebView.setLayoutParams(new LayoutParams(LayoutParams.FILL_PARENT, LayoutParams.FILL_PARENT));
+        mWebView.setDownloadListener(new DownloadListener() {
+            public void onDownloadStart(String url, String userAgent,
+                                        String contentDisposition, String mimetype,
+                                        long contentLength) {
+                Intent i = new Intent(Intent.ACTION_VIEW);
+                i.setData(Uri.parse(url));
+                startActivity(i);
+            }
+        });
         mWebView.setWebViewClient(new WebViewClient() {
-                @Override
-                public boolean shouldOverrideUrlLoading(WebView view, String url) {
-                    view.loadUrl(url);
-                    return false;
-                }
-            });
+            @Override
+            public boolean shouldOverrideUrlLoading(WebView view, String url) {
+                view.loadUrl(url);
+                return false;
+            }
+        });
+        mWebView.setWebChromeClient(new WebChromeClient() {
+            /**
+             *8 (Android 2.2) < = API < = 10 (Android 2.3) callback this method
+             */
+            private void openFileChooser(android.webkit.ValueCallback uploadMsg) {
+                mUploadCallbackBelow = uploadMsg;
+                takePhoto();
+            }
+
+            /**
+             *11 (Android 3.0) < = API < = 15 (Android 4.0.3) callback this method
+             */
+            public void openFileChooser(android.webkit.ValueCallback uploadMsg, String acceptType) {
+                openFileChooser(uploadMsg);
+            }
+
+            /**
+             *16 (Android 4.1.2) < = API < = 20 (Android 4.4w. 2) callback this method
+             */
+            public void openFileChooser(android.webkit.ValueCallback uploadMsg, String acceptType, String capture) {
+                openFileChooser(uploadMsg);
+            }
+
+            /**
+             *API > = 21 (Android 5.0.1) callback this method
+             */
+            @Override
+            public boolean onShowFileChooser(WebView webView, ValueCallback valueCallback, FileChooserParams fileChooserParams) {
+                mUploadCallbackAboveL = valueCallback;
+                takePhoto();
```

Comment:
why are we taking photo here?

Reply:
Because I found that using the camera may break the software on my old phone. But Pictures working for every time and at any phones (old and new).

In a short sentence: camera is not stable and may cause bugs.